ComparisonWithOrleans.md

```diff
@@ -0,0 +1,203 @@
+
+#### Turns
+
+  * In Orleans a Grain is scheduled in terms of non-overlapping *turns*, meaning that no two threads run that Grain’s code concurrently at any given time. This is also true for continuations formed via the usage of async–await.
+  
+  * Akka schedules Actors similarly, upon each assignment to a (single) Thread an Actor may process up to a configurable limit of messages and it is guaranteed that an Actor is running on at most one Thread at any given time. While it is possible to specify an ExecutionContext that runs continuations within the Actor’s schedule, the default executors do not respect this constraint and will typically run Future continuations concurrently with the Actor that scheduled them.
+  
+#### Persistence
+
+  * Orleans provides persistence based on snapshots that are managed by the Grain explicitly. This maps well to a database-backed model where each Grain is responsible for a conceptual “row”.
+  
+  * Akka Persistence is based upon event-sourcing, which means that the PersistentActor emits events that represent state changes to be applied. Snapshots may be used as an optimization to shorten recovery time.
+
+In both cases the persistence provider is configurable and users may choose to persist Actor state independently by other means as well. The biggest difference between snapshots and event-sourcing is that the event history of a PersistentActor contains valuable information that is lost when persisting only snapshots: each event has a well-defined meaning in the business domain and thus the event log retains its meaning as the internal representation of the Actor’s state changes over time while snapshots will become invalid and cannot be mined for additional insights later.
+
+#### Timers and Reminders
+
+  * Orleans provides transient timers that are local to a Grain activation and cease to exist when the Grain is passivated. It also provides reliable persistent reminders that will send an invocation to a Grain independently of whether it is currently active or not.
+  
+  * Akka includes only transient timers at this point, reliable reminders need to be implemented using third-party extensions like Akka Quartz.
+
+### Runtime Implementation
+
+#### Overview
+
+Both systems consist of similar core components: Orleans’ Execution corresponds to akka-actor in that it is responsible for the actual execution of Actors, Messaging corresponds to akka-remote in that it manages TCP connections between the cluster nodes, and Hosting corresponds to akka-cluster plus ClusterSharding. One difference is that akka-actor also allows purely local message sends that are optimized to local reference passing whereas Orleans invocations are always sent via the Messaging component that consults with the Hosting component as to where the recipient currently resides. Akka’s lack of indirection allows substantial performance optimizations in this regard.
```

Comment:
So there are 2 things about isolation: serialization and dispatch/context.
serialization for local calls is opt in Akka and opt out in Orleans. We agree on that.

As for dispatch, I though you wrote above: 
> One difference is that akka-actor also allows purely local message sends that are optimized to local reference passing whereas Orleans invocations are always sent via the Messaging component that consults with the Hosting component as to where the recipient currently resides.

I read from that that in Akka in the local case you inline the call to be just a regular function call by reference. Meaning the called actor will execute in the context of the calling actor, meaning breaking isolation.
If that is not the case, and you go via mailbox and context switch, than first sorry for miss-understanding.
Second,  this is exactly what Orleans is doing as well. In the local case we go via Dispatch module, of course not via Messaging component like you wrote (we make one function call: "is the destination actor already hosted on this machine". This is not considered going via the messaging component). Thus the argument about efficiency in the local case is not valid. We both have to dispatch to a different context.